README.md
# WebSocket Client for Ruby

WebSocket-EventMachine-Client is Ruby WebSocket client based on EventMachine.

- [Autobahn tests](http://imanel.github.com/websocket-ruby/autobahn/client)
- [Docs](http://rdoc.info/github/imanel/websocket-eventmachine-client/master/frames)

## Installation

``` bash
gem install websocket-eventmachine-client
```

or in Gemfile

``` ruby
gem 'websocket-eventmachine-client'
```

## Simple client example

```ruby
EM.run do

  ws = WebSocket::EventMachine::Client.connect(:uri => 'ws://localhost:8080')

  ws.onopen do
    puts "Connected"
  end

  ws.onmessage do |msg, type|
    puts "Received message: #{msg}"
  end

  ws.onclose do |code, reason|
    puts "Disconnected with status code: #{code}"
  end

  EventMachine.next_tick do
    ws.send "Hello Server!"
  end

end
```

## Options

Following options can be passed to WebSocket::EventMachine::Client initializer:

- `[String] :host` - IP or host of server to connect
- `[Integer] :port` - Port of server to connect
- `[String] :uri` - Full URI for server(optional - use instead of host/port combination)
- `[Integer] :version` - Version of WebSocket to use. Default: 13
- `[Hash] :headers` - HTTP headers to use in the handshake. Example: `{'Cookie' => 'COOKIENAME=Value'}`
- `[Boolean] :ssl` - Force SSL/TLS regardless of URI scheme or port

If you wish to connect to a UNIX socket, specify the `:host` option as the full path to the local socket and omit `:port`.

## Methods

Following methods are available for WebSocket::EventMachine::Client object:

### onopen

Called after successfully connecting.

Example:

```ruby
ws.onopen do
  puts "Client connected"
end
```

### onclose

Called after closing connection.

Parameters:

- `[Integer] code` - status code
- `[String] reason` - optional reason for closure

Example:

```ruby
ws.onclose do |code, reason|
  puts "Client disconnected with status code: #{code} and reason: #{reason}"
end
```

### onmessage

Called when client receive message.

Parameters:

- `[String] message` - content of message
- `[Symbol] type` - type is type of message(:text or :binary)

Example:

```ruby
ws.onmessage do |msg, type|
  puts "Received message: #{msg} or type: #{type}"
end
```

### onerror

Called when client discovers error.

Parameters:

- `[String] error` - error reason.

Example:

```ruby
ws.onerror do |error|
  puts "Error occured: #{error}"
end
```

### onping

Called when client receive ping request. Pong request is sent automatically.

Parameters:

- `[String] message` - message for ping request.

Example:

```ruby
ws.onping do |message|
  puts "Ping received: #{message}"
end
```

### onpong

Called when client receive pong response.

Parameters:

- `[String] message` - message for pong response.

Example:

```ruby
ws.onpong do |message|
  puts "Pong received: #{message}"
end
```

### send

Sends message to server.

Parameters:

- `[String] message` - message that should be sent to server
- `[Hash] params` - params for message(optional)
  - `[Symbol] :type` - type of message. Valid values are :text, :binary(default is :text)

Example:

```ruby
ws.send "Hello Server!"
ws.send "binary data", :type => :binary
```

### close

Closes connection and optionally send close frame to server.

Parameters:

- `[Integer] code` - code of closing, according to WebSocket specification(optional)
- `[String] data` - data to send in closing frame(optional)

Example:

```ruby
ws.close
```

### ping

Sends ping request.

Parameters:

- `[String] data` - data to send in ping request(optional)

Example:

```ruby
ws.ping 'Hi'
```

### pong

Sends pong request. Usually there should be no need to send this request, as pong responses are sent automatically by client.

Parameters:

- `[String] data` - data to send in pong request(optional)

Example:

``` ruby
ws.pong 'Hello'
```

## License

The MIT License - Copyright (c) 2012 Bernard Potocki


CHANGELOG.md
# Changelog

## 1.2.0

- support for unix domain connection

## 1.1.0

- ssl support

## 1.0.1

- ensure compatibility with em-websocket

## 1.0.0

- initial release


examples/autobahn_client.rb
require File.expand_path('../../lib/websocket-eventmachine-client', __FILE__)

require 'cgi'

EM.epoll
EM.run do

  host   = 'ws://localhost:9001'
  agent  = "WebSocket-EventMachine-Client (1.0.0)"
  cases  = 0
  skip   = []

  ws = WebSocket::EventMachine::Client.connect(:uri => "#{host}/getCaseCount")

  ws.onmessage do |msg, type|
    puts "$ Total cases to run: #{msg}"
    cases = msg.to_i
  end

  ws.onclose do

    run_case = lambda do |n|

      if n > cases
        puts "$ Requesting report"
        ws = WebSocket::EventMachine::Client.connect(:uri => "#{host}/updateReports?agent=#{CGI.escape agent}")
        ws.onclose do
          EM.stop
        end

      elsif skip.include?(n)
        EM.next_tick { run_case.call(n+1) }

      else
        ws = WebSocket::EventMachine::Client.connect(:uri => "#{host}/runCase?case=#{n}&agent=#{CGI.escape agent}")

        ws.onmessage do |msg, type|
          ws.send(msg, :type => type)
        end

        ws.onclose do |msg|
          EM.add_timer(0.1) { run_case.call(n + 1) }
        end
      end
    end

    run_case.call(1)
  end

end


lib/websocket/eventmachine/client.rb
require 'websocket-eventmachine-base'
require 'uri'

module WebSocket
  module EventMachine

    # WebSocket Client (using EventMachine)
    # @example
    #   ws = WebSocket::EventMachine::Client.connect(:host => "0.0.0.0", :port => 8080)
    #   ws.onmessage { |msg| ws.send "Pong: #{msg}" }
    #   ws.send "data"
    class Client < Base

      ###########
      ### API ###
      ###########

      # Connect to websocket server
      # @param args [Hash] The request arguments
      # @option args [String] :host The host IP/DNS name
      # @option args [Integer] :port The port to connect too(default = 80)
      # @option args [String] :uri Full URI for server(optional - use instead of host/port combination)
      # @option args [Integer] :version Version of protocol to use(default = 13)
      # @option args [Hash] :headers HTTP headers to use in the handshake
      # @option args [Boolean] :ssl Force SSL/TLS connection
      def self.connect(args = {})
        host = nil
        port = nil
        if args[:uri]
          uri = URI.parse(args[:uri])
          host = uri.host
          port = uri.port
          args[:ssl] = true if uri.scheme == 'wss'
        end
        host = args[:host] if args[:host]
        port = args[:port] if args[:port]
        if args[:ssl]
          port ||= 443
        else
          port ||= 80
        end

        ::EventMachine.connect host, port, self, args
      end
      
      # Make a websocket connection to a UNIX-domain socket.
      # @param socketname [String] Unix domain socket (local fully-qualified path)
      # @param args [Hash] Arguments for connection
      # @option args [Integer] :version Version of protocol to use(default = 13)
      # @option args [Hash] :headers HTTP headers to use in the handshake
      def self.connect_unix_domain(socketname, args = {})
        fail ArgumentError, 'invalid socket' unless File.socket?(socketname)
        args[:host] ||= 'localhost'
        ::EventMachine.connect_unix_domain socketname, self, args
      end

      # Initialize connection
      # @param args [Hash] Arguments for connection
      # @option args [String] :host The host IP/DNS name
      # @option args [Integer] :port The port to connect too(default = 80)
      # @option args [Integer] :version Version of protocol to use(default = 13)
      # @option args [Hash] :headers HTTP headers to use in the handshake
      # @option args [Boolean] :ssl Force SSL/TLS connection
      def initialize(args)
        @args = args
      end

      ############################
      ### EventMachine methods ###
      ############################

      # Called after initialize of connection, but before connecting to server
      # Eventmachine internal
      # @private
      def post_init
        @state = :connecting
        @handshake = ::WebSocket::Handshake::Client.new(@args)
      end

      # Called by EventMachine after connecting.
      # Sends handshake to server or starts SSL/TLS
      # Eventmachine internal
      # @private
      def connection_completed
        if @args[:ssl]
          start_tls
        else
          send(@handshake.to_s, :type => :plain)
        end
      end

      # Called by EventMachine after SSL/TLS handshake.
      # Sends websocket handshake
      # Eventmachine internal
      # @private
      def ssl_handshake_completed
        send(@handshake.to_s, :type => :plain)
      end

      private

      def incoming_frame
        ::WebSocket::Frame::Incoming::Client
      end

      def outgoing_frame
        ::WebSocket::Frame::Outgoing::Client
      end

      public

      #########################
      ### Inherited methods ###
      #########################

      # Called when connection is opened.
      # No parameters are passed to block
      def onopen(&blk); super; end

      # Called when connection is closed.
      # No parameters are passed to block
      def onclose(&blk); super; end

      # Called when error occurs.
      # One parameter passed to block:
      #   error - string with error message
      def onerror(&blk); super; end

      # Called when message is received.
      # Two parameters passed to block:
      #   message - string with received message
      #   type - type of message. Valid values are :text and :binary
      def onmessage(&blk); super; end

      # Called when ping message is received
      # One parameter passed to block:
      #   message - string with ping message
      def onping(&blk); super; end

      # Called when pong message is received
      # One parameter passed to block:
      #   message - string with pong message
      def onpong(&blk); super; end

      # Send data
      # @param data [String] Data to send
      # @param args [Hash] Arguments for send
      # @option args [String] :type Type of frame to send - available types are "text", "binary", "ping", "pong" and "close"
      # @option args [Integer] :code Code for close frame
      # @return [Boolean] true if data was send, otherwise call on_error if needed
      def send(data, args = {}); super; end

      # Close connection
      # @return [Boolean] true if connection is closed immediately, false if waiting for other side to close connection
      def close(code = 1000, data = nil); super; end

      # Send ping message
      # @return [Boolean] false if protocol version is not supporting ping requests
      def ping(data = ''); super; end

      # Send pong message
      # @return [Boolean] false if protocol version is not supporting pong requests
      def pong(data = ''); super; end

    end
  end
end
